src/core/namespace.c

```diff
@@ -3155,6 +3156,107 @@ bool ns_type_supported(NamespaceType type) {
         return access(ns_proc, F_OK) == 0;
 }
 
+int refresh_extensions_in_namespace(
+                const PidRef *target,
+                const char *hierarchy_env,
+                const NamespaceParameters *p) {
+        const char *overlay_prefix = "/run/systemd/mount-rootfs";
+        _cleanup_(mount_list_done) MountList ml = {};
+        _cleanup_free_ char *extension_dir = NULL;
+        _cleanup_strv_free_ char **hierarchies = NULL;
+        int r;
+
+        assert(target);
+        assert(hierarchy_env);
+        assert(p);
+
+        log_debug("Refreshing extensions in-namespace for hierarchy '%s'", hierarchy_env);
+
+        extension_dir = strjoin(p->private_namespace_dir, "/unit-extensions");
+        if (!extension_dir)
+                return -ENOMEM;
+
+        r = parse_env_extension_hierarchies(&hierarchies, hierarchy_env);
+        if (r < 0)
+                return r;
+
+        r = append_extensions(
+                        &ml,
+                        overlay_prefix,
+                        p->private_namespace_dir,
+                        hierarchies,
+                        p->extension_images,
+                        p->n_extension_images,
+                        p->extension_directories);
+        if (r < 0)
+                return r;
+        if (ml.n_mounts == 0)
+                return 0;
+
+        (void) mkdir_p(overlay_prefix, 0755);
+
+        /* This is effectively two rounds, since all the extensions come before
+         * overlays (setup_namespace() similarly relies on this property).
+         *
+         * (1) First, set up all the extension mounts in the system, which are
```

Comment:
This will create new mounts in the root namespace, which is not great as it will create confusion and noise in the mounts list. So use safe_fork here, with FORK_NEW_MOUNTNS and FORK_WAIT, and run the loop in the child, that way they will all be hidden., and you can drop the manual umounts too as they'll disappear once the child process exits

Reply:
Switched to forking and dropped the unmounting code, PTAL.